fabric_generator/utilities.py

```diff
@@ -0,0 +1,1140 @@
+                # Handle tile attributes depending on their label
+                if wire[0] == "MATRIX":
+                    vhdlLoc = wire[1]
+                    csvLoc = vhdlLoc[:-4:] + "csv"
+                    cTile.matrixFileName = csvLoc
+                    try:
+                        csvFile = RemoveComments(
+                            [i.strip('\n').split(',') for i in open(csvLoc)])
+                        cTile.pips = findPipList(csvFile)
+
+                        cTile.pipMuxes_MapSourceToSinks = findPipList(
+                            csvFile, returnDict=True, mapSourceToSinks=True)
+                        cTile.pipMuxes_MapSinkToSources = findPipList(
+                            csvFile, returnDict=True, mapSourceToSinks=False)
+
+                    except:
+                        raise Exception("CSV File not found.")
+
+                if wire[0] == "BEL":
+                    belHasClockInput = False
+                    try:
+                        ports = GetComponentPortsFromFile(wire[1])
+
+                        # We also want to check whether the component has a clock input
+                        # Get all external (routed to top) ports
+                        externalPorts = (GetComponentPortsFromFile(
+                            wire[1], port="external"))
+                        for port in externalPorts:
+                            # Get port name
+                            PortName = re.sub('\:.*', '', port)
+                            substitutions = {" ": "", "\t": ""}  # Strip
+                            PortName = (replace(PortName, substitutions))
+                            if PortName == "UserCLK":  # And if UserCLK is in there then we have a clock input
+                                belHasClockInput = True
+
+                    except:
+                        raise Exception(f"{wire[1]} file for BEL not found")
+
+                    if len(wire) > 2:
+                        prefix = wire[2]
+                    else:
+                        prefix = ""
+                    nports = []
+                    inputPorts = []
+                    outputPorts = []
+
+                    for port in ports[0]:
+                        nports.append(
+                            prefix + re.sub(" *\(.*\) *", "", str(port)))
+                        # Also add to distinct input/output lists
+                        inputPorts.append(
+                            prefix + re.sub(" *\(.*\) *", "", str(port)))
+                    for port in ports[1]:
+                        nports.append(
+                            prefix + re.sub(" *\(.*\) *", "", str(port)))
+                        outputPorts.append(
+                            prefix + re.sub(" *\(.*\) *", "", str(port)))
+                    cTile.belPorts.update(nports)
+
+                    belListWithIO.append(
+                        [wire[1][0:-5:], prefix, inputPorts, outputPorts, belHasClockInput])
```

Comment:
hmm, this hardcoded -5 looks suspicious, what is it intended for?

Reply:
seems like this was intended for stripping off a ".vhdl" extension, but it doesn't work any more now that these are full paths and you also need to strip off upto and including the "/", so this logic needs to be fixed. It would also always break for a Verilog bel or a ".vhd" extension...